gh-127787: refactor helpers for `PyUnicodeErrorObject` internal interface

picnixz opened the pull request:
- Unify `get_unicode` and `get_string` in a single function.
- Allow to retrieve the underlying `object` attribute, its size and its start and end indices in one round.
- Use a common implementation for the following functions:

  - `PyUnicode{Decode,Encode}Error_GetEncoding`
  - `PyUnicode{Decode,Encode,Translate}Error_GetObject`
  - `PyUnicode{Decode,Encode,Translate}Error_{Get,Set}Reason`
  - `PyUnicode{Decode,Encode,Translate}Error_{Get,Set}{Start,End}`

Note that there are some cosmetic changes here and there (in the naming of parameters) but these are essentially in prevision of https://github.com/python/cpython/pull/127694 in order to reduce the conflicts I'll need to solve (there will be conflicts probably but ideally, I want them to be minimal).

I've moved all helpers before the public API. I could move them inbetween but I felt that it's cleaner that way (it also allowed me to put double blank lines between functions a bit more easily).


<!-- gh-issue-number: gh-127787 -->
* Issue: gh-127787
<!-- /gh-issue-number -->

picnixz committed:
Refactor helpers for `PyUnicodeErrorObject`.

- Unify `get_unicode` and `get_string` in a single function.
- Allow to retrieve the underlying `object` attribute and its
  size in one round.
- Use a common implementation for the following functions:

  - `PyUnicode{Decode,Encode}Error_GetEncoding`
  - `PyUnicode{Decode,Encode,Translate}Error_GetObject`
  - `PyUnicode{Decode,Encode,Translate}Error_{Get,Set}Reason`
  - `PyUnicode{Decode,Encode,Translate}Error_{Get,Set}{Start,End}`

picnixz committed:
put comment section headers

picnixz committed:
add comments

picnixz committed:
simpler checks

picnixz committed:
fix tests

picnixz committed:
unify even more the interface using a generic getter

picnixz committed:
remove useless `consistent` parameter

picnixz committed:
Simplify call

picnixz committed:
Merge branch 'main' into feat/exc/unicode-error-refactor-127787

picnixz committed:
remove unused function

picnixz committed:
remove un-necessary macros

encukou commented on Objects/exceptions.c:
```diff
@@ -2667,47 +2667,140 @@ SimpleExtendsException(PyExc_Exception, ValueError,
-        PyErr_Format(PyExc_TypeError, "%.200s attribute must be bytes", name);
+    if (!(as_bytes ? PyBytes_Check(attr) : PyUnicode_Check(attr))) {
+        PyErr_Format(PyExc_TypeError,
+                     "%.200s attribute must be %s",
```
While you're here: these limits haven't been needed for a long time now.
```suggestion
                     "%s attribute must be %s",
```

encukou commented on Objects/exceptions.c:
```diff
@@ -2667,47 +2667,140 @@ SimpleExtendsException(PyExc_Exception, ValueError,
+/*
+ * Return the underlying (str) 'encoding' attribute of a Unicode Error object.
+ *
+ * The caller is responsible to ensure that 'self' is a PyUnicodeErrorObject.
```
I'd prefer an `assert` over a “The caller is responsible to ensure...” comment. To a human reader, they should be equivalent.

picnixz replied:
The `assert` is actually inside the `_CAST` macro. It's just to document that this function would crash otherwise. The alternative is to remove the assert inside the CAST macro and make it an explicit one though that would add lines.

picnixz committed:
Update Objects/exceptions.c

Co-authored-by: Petr Viktorin <encukou@gmail.com>

picnixz committed:
style update

picnixz committed:
use macro for repeated names to avoid typos

picnixz committed:
specialize _PyUnicodeError_GetParams for start and end attributes

This is typically useful for future refactorization and to
be able to write lines below 80 characters. This also helps
avoiding having to remember where to place the NULL arguments.

picnixz committed:
update comments

picnixz committed:
put 2 blank lines before sections

picnixz committed:
add NULL assertions to avoid obscure segmentation faults

encukou commented on Objects/exceptions.c:
```diff
@@ -2667,47 +2667,170 @@ SimpleExtendsException(PyExc_Exception, ValueError,
+
+
+/*
+ * Check that 'self' is of a Unicode Error object.
```
`UnicodeError` is the Python class name, no need for a space.

Consider pointing out the lax type check; it's a surprising detail:
```suggestion
 * Check that 'self' is of a UnicodeError object.
 *
 * `expect_type` is the name of the expected type, and only used for
 * error reporting.
 * (As an implementation detail, each of the the `PyUnicode*Error_*`
 * functions actually allows *any* `UnicodeError` subclass
 * as `self`.)
 * Use one of the `Py_UNICODE_*_ERROR_NAME` macros to avoid typos.
```

picnixz replied:
> UnicodeError is the Python class name, no need for a space.

I should have actually written `Unicode error object`, namely an error object for Unicode stuff. I'll also update the other occurrences of Unicode Error object (I'll use UnicodeError object)

encukou commented on Objects/exceptions.c:
```diff
@@ -2748,286 +2872,290 @@ unicode_error_adjust_end(Py_ssize_t end, Py_ssize_t objlen)
+ *     as_bytes     Indicate whether the underlying 'object' is a bytes object.
+ *
+ * The 'obj', 'objlen', 'start' and 'end' parameters may be NULL
+ * to indicate that the parameter does not need to be stored.
```
Perhaps group these?
```suggestion
 * Output parameters:
 *
 *     obj          The 'object' attribute.
 *     objlen       The length of 'object'.
 *     start        The clipped 'start' attribute.
 *     end          The clipped 'end' attribute.
 *
 * Any of these may be NULL to indicate that the value does
 * not need to be stored.
 *
 * Input parameter:
 *     as_bytes     If 1, the error's 'object' attribute must be a bytes object
 *                    (i.e. the call is for a `UnicodeDecodeError`).
 *                  If 0, 'object' must be a string.
 *                  Raise TypeError if it has a wrong type.
 *
```

picnixz replied:
That would probably be better for the reader!

encukou commented on Objects/exceptions.c:
```diff
@@ -2667,47 +2667,170 @@ SimpleExtendsException(PyExc_Exception, ValueError,
+ * Return the underlying (str) 'encoding' attribute of a Unicode Error object.
+ *
+ * The caller is responsible to ensure that 'self' is a PyUnicodeErrorObject,
+ * although this condition is verified by this function on DEBUG builds.
```
For internal functions, where the reader can see the code with the `_CAST`, repeating these 2 lines every time seems too verbose.
(The comment helps if you forget that `_CAST` contains an `assert`, like I did, but it's not worth constant reminders :)

picnixz replied:
I've added a top-level (section) comment and removed all occurrences of those comments, except when a new section is created.

encukou reviewed (commented):
LGTM! Just a few comment nitpicks left.

picnixz committed:
update comments

picnixz committed:
fixup comment

picnixz committed:
fixup comment

encukou reviewed (approved):
Thank you!
